notebooks/11-Advanced-Pandas.ipynb

```diff
@@ -0,0 +1,2864 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 24,
+   "id": "7e3bf483-3395-4eae-a588-1d7a5d7b52a4",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "fragment"
+    },
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "pool.close()\n",
+    "pool.join()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "96560f76-cded-4aff-8df8-2237b470c8c8",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    },
+    "tags": []
+   },
+   "source": [
+    "Although this is cumbersome, we could convert it to a reusable function:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 26,
+   "id": "6d060d59-6266-47e3-86ed-34a1660237ee",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "fragment"
+    },
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "def parallel_apply(\n",
+    "    df, func, cores\n",
+    "):\n",
+    "    pool = Pool(cores)\n",
+    "    splits = np.array_split(df, cores)\n",
+    "    split_work = pool.map(\n",
+    "        func, splits\n",
+    "    )\n",
+    "    df = pd.concat(split_work)\n",
+    "    pool.close()\n",
+    "    pool.join()\n",
+    "    \n",
+    "    return df"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "583c43c1-698a-4566-9516-056e7f1a8602",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    },
+    "tags": []
+   },
+   "source": [
+    "Something to note here is how much extra work we're doing to apply a function in parallel:\n",
+    "\n",
+    "* Split the data frame into N copies\n",
+    "* Fork 4 new Python processes\n",
+    "* Run the function N times\n",
+    "* Concatenate the results back into one data frame\n",
+    "* Clean up"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "033d7990-8f94-4461-b7e7-6be899bb4f3e",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "fragment"
+    },
+    "tags": []
+   },
+   "source": [
+    "In other words: you should only do this if the code you're parallelizing is sufficiently time-intensive to make it worthwhile. Otherwise you could easily end up taking more time than you would if you'd remained single-threaded."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c5027eb8-bc68-4b4f-bb27-3b0f3f1a1cbc",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    },
+    "tags": []
+   },
+   "source": [
+    "**Note:** In general, if you have a substantial amount of data and wish to leverage multiple cores or CPUs, you might be better served investigating a library that is built from the ground up to allow for distributed processing. Dask and PySpark are two possible alternatives in this space."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "932e9eec-e941-455b-9824-42d1827e55bd",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    },
+    "tags": []
+   },
+   "source": [
+    "## Joining Data Frames"
```

Comment:
Should we move the joining section earlier? I would think we want to move from more common to less common use cases (with some flexibility for naturally building concepts), and joining data frames should be more relevant/common that parallelization.